packages/bot-skeleton/src/services/tradeEngine/trade/index.js

```diff
@@ -107,45 +107,46 @@ export default class TradeEngine extends Balance(Purchase(Sell(OpenContract(Prop
+            // "proposal_open_contract" which would make the bot run forever. When there's a "sell"
+            // event, wait a couple seconds for the API to give us the correct "proposal_open_contract"
+            // response, if there's none after x seconds. Send an explicit request, which _should_
+            // solve the issue. This is a backup!
+            this.api.onMessage().subscribe(({ data }) => {
+                if (data.msg_type === 'transaction' && data.transaction.action === 'sell') {
+                    this.transaction_recovery_timeout = setTimeout(() => {
+                        const { contract } = this.data;
+                        const is_same_contract = contract.contract_id === data.transaction.contract_id;
+                        const is_open_contract = contract.status === 'open';
+                        if (is_same_contract && is_open_contract) {
+                            doUntilDone(() => {
+                                this.api.send({ proposal_open_contract: 1, contract_id: contract.contract_id });
+                            }, ['PriceMoved']);
+                        }
+                    }, 1000);
                 }
-
-                // Try to recover from a situation where API doesn't give us a correct response on
-                // "proposal_open_contract" which would make the bot run forever. When there's a "sell"
-                // event, wait a couple seconds for the API to give us the correct "proposal_open_contract"
-                // response, if there's none after x seconds. Send an explicit request, which _should_
-                // solve the issue. This is a backup!
-                this.listen('transaction', ({ transaction }) => {
-                    if (transaction.action === 'sell') {
-                        this.transaction_recovery_timeout = setTimeout(() => {
-                            const { contract } = this.data;
-                            const is_same_contract = contract.contract_id === transaction.contract_id;
-                            const is_open_contract = contract.status === 'open';
-
-                            if (is_same_contract && is_open_contract) {
-                                doUntilDone(() => this.api.getContractInfo(contract.contract_id));
-                            }
-                        }, 1000);
+            });
+            this.api.onMessage().subscribe(async ({ data }) => {
+                if (data.msg_type === 'authorize') {
+                    this.accountInfo = data;
+                    this.token = token;
+
+                    // Only subscribe to balance in browser, not for tests.
+                    if (document) {
+                        const balance_request = await this.api.send({ balance: 1, subscribe: 1 });
+                        this.balance_request = Number(balance_request.balance.balance);
+                        resolve();
+                    } else {
+                        resolve();
```

Comment:
Do we need an else if we're only using it to resolve the Promise?

Reply:
Yes, We have a promise and we need to resolve() although here resolve empty. but I've updated it and has one resolve